### PR TITLE
Display ZKTokens and Balances

### PR DESCRIPTION
**Pull request by owonwo**
Feature: Display ZKTokens and Balances.

### CODE REVIEW

File: packages/zk-transfer-web/src/App.tsx [hunk 1]
```diff
@@ -16,24 +18,42 @@ type AppProp = {
...
     if (!chainId) {
       demo.clientInit()
-        .then(() => {
+        .then(async () => {
+          console.groupCollapsed('Client Initialized');
+          const api = await makeApi(demo);
+          //@ts-ignore
+          window.api = api; window.demo = demo;
+          api.zkTokens.mapEntries(([tokenName, token]) => {
+            const shortTokenName = getZkTokenName(tokenName);
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/App.tsx]: again, please use `tradeSymbol` here and everywhere else

File: packages/zk-transfer-web/src/App.tsx [hunk 2]
```diff
@@ -16,24 +18,42 @@ type AppProp = {
...
     const { chainId } = state;
 
     if (!chainId) {
       demo.clientInit()
-        .then(() => {
+        .then(async () => {
+          console.groupCollapsed('Client Initialized');
+          const api = await makeApi(demo);
+          //@ts-ignore
+          window.api = api; window.demo = demo;
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/App.tsx]: I'm not sure this is the right place to put it.
Doesn't webpack already populate `window.api` and `window.demo` ?

File: packages/zk-transfer-web/src/index.tsx [hunk 3]
```diff
@@ -3,12 +3,15 @@ import ReactDOM from "react-dom";
...
 import App from "./App";
+import Offline from './components/Offline';
 import * as serviceWorker from "./serviceWorker";
 
 document.addEventListener('DOMContentLoaded', async () => {
     const { demo } : any = window;
-    
-    ReactDOM.render(<App demo={demo}/>, document.getElementById("root"));        
+
+    const Component = (demo) ? <App demo={demo} /> : <Offline />;
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/index.tsx]: Nice! It seems this deals with the problem where HTTP(S) requests from custom ports like 7001 sometimes get filtered out by firewalls.

However, in your case, without the VPN, the user is not technically offline. Otherwise they would never be able to load the page in the first place! We'll leave it for a future issue how to notify the user in that case, or to only use standard ports like 80 and 443 for both `ETH_URL` and `DB_URL`.

File: packages/zk-transfer-web/src/components/Token.tsx [hunk 4]
```diff
@@ -1,31 +1,55 @@
...
+  const pendingValues: NoteValue[] = notes.map(getNoteValue);
+  return Promise.all(Array.from(pendingValues));
+}
+
 const TokenInput = (props: TokenPropType) => {
-  const editMode = props.canEdit;
-  const [digit, setDigit] = useState(props.firstValue);
   const input = useRef(null);
+  const editMode = props.canEdit;
+  const [digit, setDigit] = useState(0);
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/components/Token.tsx]: what is `digit`?
Why save the `fetched` state?
I thought that React suspense and fetchers handled caching values internally.

File: packages/zk-transfer-web/src/components/Token.tsx [hunk 5]
```diff
@@ -64,6 +88,13 @@ const TokenInput = (props: TokenPropType) => {
   );
 };
 
+export const Skeleton = () => {
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/components/Token.tsx]: What is the need for a Skeleton react component?

File: packages/zk-transfer-web/src/libs/types.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
+export type ERCToken = Map<any, any>;
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/libs/types.ts]: This should more specifically be called an `ERC20Token` as there are many ERCs out there, standards for describing other things besides tokens. What are the keys and values? Can we be more specific about the types?

File: packages/zk-transfer-web/src/util.ts [hunk 7]
```diff
@@ -1,15 +1,44 @@
...
+  const replaceFn = (tokn: any, $1: any, $2: any) =>
+    capitalize($1) + " " + capitalize($2);
   return screenNameGenerator.getRandomName().replace(regex, replaceFn);
-}
\ No newline at end of file
+};
+
+export const getZkTokenName = (token: string | number) => {
+  const result = /ZkAssetTradeable-deploy(.+)/.exec(String(token));
+  if (result) {
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/util.ts]: how about
`if (result.length > 1)`
or
`if (result.length === 1)`

File: packages/zk-transfer-web/src/libs/types.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+export interface Note extends Map<string, any> {
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/libs/types.ts]: Interesting. Did you have to generate this list by hand?
If so, that seems very labor-intensive, and I'm sorry you had to go through that.
If it appears in the future that you'll have to do something very repetitive like this, check with me first.

Given your experience here, you might be able to help me simplify or re-write a different part of the project, called a typed transform pipeline.

File: packages/zk-transfer-web/src/components/Token.tsx [hunk 9]
```diff
@@ -136,4 +167,4 @@ export const TokenCard = (props: any) => {
   );
 };
 
-export default TokenInput;
+export default React.memo(TokenInput);
```
[the remark this answers is not in the document]
Comment (owonwo) [reply packages/zk-transfer-web/src/components/Token.tsx]: The `<TokenInput>` component renders everytime the cursor overs on them. So the `React.memo` wrapper will memoize (cache) the component. This will reduce the number of times the component gets to re-render

File: packages/zk-transfer-web/src/components/Token.tsx [hunk 10]
```diff
@@ -1,31 +1,55 @@
...
   useEffect(() => {
     canEdit((a: any) => {
-      // console.log('The input at first', a);
       if (a) a.current.focus();
     })(input);
-  });
+    if (!fetched && props.notes) {
+       console.log('Fetching balances for ', props.name);
+       fetchValue(props.notes)
+        .then(values => setDigit(values.reduce(getSum, 0)));
```
[the remark this answers is not in the document]
Comment (owonwo) [reply packages/zk-transfer-web/src/components/Token.tsx]: It's okay. Because i've used it only one time.

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 5 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/components/Token.tsx]: That's the animating placeholder that shows up when the Tokens are being fetched.

File: packages/zk-transfer-web/src/libs/types.ts (hunk 6 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/libs/types.ts]: I haven't see the structure of the ERC20Token. So that's why i give it a primitive type. I'll replace it once i see it's structure.

File: packages/zk-transfer-web/src/util.ts (hunk 7 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/util.ts]: I meant to use `.test` not `.exec` I'll change it.

File: packages/zk-transfer-web/src/libs/types.ts (hunk 8 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/libs/types.ts]: I did generate it by hand. Just to stop the errors from typescript.

File: packages/zk-transfer-web/src/util.ts (hunk 7 above)
Comment (cryptogoth) [reply packages/zk-transfer-web/src/util.ts]: ok. What about line 21?
If result is non-null, but has zero elements, or more than 1 element,
it is likely an error.

File: packages/zk-transfer-web/src/libs/types.ts (hunk 6 above)
Comment (cryptogoth) [reply packages/zk-transfer-web/src/libs/types.ts]: Sure. It is exactly the same as the ZKToken,
they are all smart contracts that are deployed, and associated with a trading symbol like "AAA"

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (cryptogoth) [reply packages/zk-transfer-web/src/components/Token.tsx]: ok, please make the change, thanks.

File: packages/zk-transfer-web/src/libs/types.ts (hunk 8 above)
Comment (cryptogoth) [reply packages/zk-transfer-web/src/libs/types.ts]: Thanks for doing that this time!

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 4 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/components/Token.tsx]: So, i wrapped the logic in a Custom Hook insted of having to use the React suspense (since it's still experimental).

File: packages/zk-transfer-web/src/util.ts (hunk 7 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/util.ts]: Okay.

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (owonwo) [reply packages/zk-transfer-web/src/components/Token.tsx]: When i try this `parseInt(v, 16)`, i get `33` instead of `22` @cryptogoth

File: packages/zk-transfer-web/src/App.tsx [hunk 11]
```diff
@@ -16,24 +18,42 @@ type AppProp = {
...
   });
 
   useEffect(() => {
     const { chainId } = state;
 
     if (!chainId) {
       demo.clientInit()
-        .then(() => {
+        .then(async () => {
+          console.groupCollapsed('Client Initialized');
```
[the remark this answers is not in the document]
Comment (owonwo) [reply packages/zk-transfer-web/src/App.tsx]: I guess we learn everday. :smile:

File: packages/zk-transfer-web/src/App.tsx [hunk 12]
```diff
@@ -1,57 +1,52 @@
...
 
 type AppProp = {
   demo: any
 };
 
 function App({ demo } : AppProp) {
-  const [state, setState] = useState({
-    chainId: demo.chainId,
+  const [state, setState]: [any, any] = useState({
+    chainId: null,
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/App.tsx]: good catch, this will always be null when the component first loads.

File: packages/zk-transfer-web/src/libs/api.ts [hunk 13]
```diff
@@ -31,24 +38,23 @@ export const makeApi = async (demo: any): Promise<apiObject> => {
...
+  const deploys: TokenMap = await bm.getDeploys();
   const ace = (await demo.contract.createContract("ACE")).getInstance();
 
-  const filterToken = (list: TokenList<Token>) => (regex: RegExp) => {
+  const filterToken = (list: TokenMap) => (regex: RegExp) => {
     return list.filter((_, name) => name.match(regex));
   };
 
-  const securities = filterToken(deploys)(/deploy[A-Z][A-Z][A-Z]/);
+  const securities = filterToken(deploys)(/deploy[A-Z]{3}/);
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/libs/api.ts]: 👍

File: packages/zk-transfer-web/src/libs/api.ts [hunk 14]
```diff
@@ -31,24 +38,23 @@ export const makeApi = async (demo: any): Promise<apiObject> => {
   const demoAztec = window.aztec;
   const chainId = await eth.net_version();
   const bm = await demo.contract.createBM({ chainId, autoConfig: true });
-  const deploys: TokenList<Token> = await bm.getDeploys();
+  const deploys: TokenMap = await bm.getDeploys();
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/libs/api.ts]: A better name for this type is `TokenDeploys` here and everywhere else :)
please change

File: packages/zk-transfer-web/src/components/Token.tsx [hunk 15]
```diff
@@ -57,15 +71,15 @@ const TokenInput = (props: TokenPropType) => {
       {props.children}
       <input
         ref={input}
-        value={digit}
+        value={balance}
         disabled={!props.canEdit}
-        onChange={canEdit((e: any) => setDigit(e.target.value))}
+        onChange={canEdit((e: any) => {})}
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/components/Token.tsx]: why this change?

File: packages/zk-transfer-web/src/App.tsx [hunk 16]
```diff
@@ -1,57 +1,52 @@
...
+    if (!state.chainId)
       demo.clientInit()
         .then(async () => {
           console.groupCollapsed('Client Initialized');
-          const api = await makeApi(demo);
-          api.zkTokens.mapEntries(([tokenName, token]) => {
-            const shortTokenName = getZkTokenName(tokenName);
-            const tokenAddress: string = token.get("deployAddress");
+          const { zkTokens, thisAddressNotes } = await makeApi(demo);
+          const tokenAndNotes = zkTokens.mapEntries(([ tokenName, token]) => {
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/App.tsx]: a better name for `tokenAndNotes` is `tradeSymbolToNotes` or `notesMap`,
with a comment that tells us what the key is and the value is. like
```
// map of tradeSymbol (a string) to a NoteList

File: packages/zk-transfer-web/src/util.ts [hunk 17]
```diff
@@ -17,11 +17,12 @@ export const generateScreenName = () => {
 };
 
 export const getZkTokenName = (token: string | number) => {
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/util.ts]: `token` here is better called `tokenDeployName`, which elsewhere in the code we understand to mean a name that looks like `<SmartContractName>-deploy<deployId>`.

Comment (cryptogoth) [reply packages/zk-transfer-web/src/util.ts]: All this method name should be called `getZkTradeSymbol`, it's more accurate. Thanks

File: packages/zk-transfer-web/src/libs/types.ts [hunk 18]
```diff
@@ -27,27 +33,17 @@ export interface ZkToken extends Map<string, any> {
...
-  mapEntries<U>(f: (first: KeyValuePair<Note>, key: string) => U): NoteList;
-}
+export type TokenMap = Map<TokenAddress, Token>;
 
-export interface BN {
-  add(address: string): Number;
-}
+export type TokenNotesPair = [TokenAddress, NoteList];
+
+export type TokenNotesMap = Map<TokenAddress, NoteList>;
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/libs/types.ts]: I'd prefer this type to be `TokenAddressToNoteList` as it tells you exactly
what the types and meaning of the keys and values are.

File: packages/zk-transfer-web/src/components/Token.tsx [hunk 19]
```diff
@@ -1,33 +1,55 @@
...
+        setError(err);
+      }
+  }
+
+  useEffect(() => {
+    fetchNotes(notes);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [notes]);
+
+  return [ balance, error ];
```
Comment (cryptogoth) [inline packages/zk-transfer-web/src/components/Token.tsx]: I prefer `{ balance, error }` here,  so that when you destructure it below as
`{ balance}`, the order doesn't matter.
What do you think?

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 15 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/components/Token.tsx]: It's not needed yet!

File: packages/zk-transfer-web/src/util.ts (hunk 17 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/util.ts]: Done!

File: packages/zk-transfer-web/src/App.tsx (hunk 16 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/App.tsx]: I didn't add the comment because Typescript can tell the type for the `key` and `value`

File: packages/zk-transfer-web/src/libs/api.ts (hunk 14 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/libs/api.ts]: Done!

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 19 above)
Comment (owonwo) [reply packages/zk-transfer-web/src/components/Token.tsx]: I thought of too. Good one

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 15 above)
Comment (cryptogoth) [reply packages/zk-transfer-web/src/components/Token.tsx]: the suspense!

File: packages/zk-transfer-web/src/components/Token.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (cryptogoth) [reply packages/zk-transfer-web/src/components/Token.tsx]: You mean 34? That's the decimal value for the hex value 22.
However, you're parsing correctly in your code, since minting of a 10-value note shows up correct as a (decimal) 10 balance, not as 16 (which would be the value if we converted it as `parseInt('10', 16)`

**Comment by cryptogoth** [comment]
This looks great!
Thanks for taking your time to understand the task, and working through the code review and co-learning with me.
